src/cdomains/arrayDomain.ml

```diff
@@ -941,7 +1003,762 @@ struct
+        NotNull
+        (* ... else return Top *)
+      else
+        Top
+    (* if there is no maximum size *)
+    | Some max_i, None when Z.geq max_i Z.zero ->
+      (* ... and maximum value in index interval < minimal size, return Null if all numbers in index interval are in must_nulls_set *)
+      if Z.lt max_i min_size && all_indexes_must_null min_i max_i then
+        Null
+        (* ... return NotNull if no number in index interval is in may_nulls_set *)
+      else if not (MayNulls.exists (fun x -> Z.geq x min_i && Z.leq x max_i) may_nulls_set) then
+        NotNull
+      else
+        Top
+    | Some max_i, Some max_size when Z.geq max_i Z.zero ->
+      (* if maximum value in index interval < minimal size, return Null if all numbers in index interval are in must_nulls_set *)
+      if Z.lt max_i min_size && all_indexes_must_null min_i max_i then
+        Null
+        (* if maximum value in index interval < maximal size, return NotNull if no number in index interval is in may_nulls_set *)
+      else if Z.lt max_i max_size && not (MayNulls.exists (fun x -> Z.geq x min_i && Z.leq x max_i) may_nulls_set) then
+        NotNull
+      else
+        Top
+    (* if maximum number in interval is invalid, i.e. negative, return Top of value *)
+    | _ -> Top
+
+  let set (ask: VDQ.t) (must_nulls_set, may_nulls_set, size) (e, i) v =
+    let rec add_indexes i max may_nulls_set =
+      if Z.gt i max then
+        may_nulls_set
+      else
+        add_indexes (Z.succ i) max (MayNulls.add i may_nulls_set) in
+    let min interval = match Idx.minimal interval with
+      | Some min_num when Z.geq min_num Z.zero -> min_num
+      | _ -> Z.zero in (* assume worst case minimal natural number *)
+
+    let min_size = min size in
+    let min_i = min i in
+    let max_i = idx_maximal i in
+
+    let set_exact i =
+      match idx_maximal size with
+      (* if size has no upper limit *)
+      | None -> 
+        (* ... and value <> null, remove i from must_nulls_set and also from may_nulls_set if not top *)
+        if Val.is_not_null v && not (MayNulls.is_top may_nulls_set) then
+          (must_nulls_remove i must_nulls_set min_size, MayNulls.remove i may_nulls_set, size)
+        else if Val.is_not_null v then
+          (must_nulls_remove i must_nulls_set min_size, may_nulls_set, size)
+          (* ..., i < minimal size and value = null, add i to must_nulls_set and may_nulls_set *)
+        else if Z.lt i min_size && Val.is_null v then
+          (MustNulls.add i must_nulls_set, MayNulls.add i may_nulls_set, size)
+          (* ..., i >= minimal size and value = null, add i only to may_nulls_set *)
+        else if Val.is_null v then
+          (must_nulls_set, MayNulls.add i may_nulls_set, size)
+          (* ... and value unknown, remove i from must_nulls_set and add it to may_nulls_set *)
+        else
+          (must_nulls_remove i must_nulls_set min_size, MayNulls.add i may_nulls_set, size)
```

Comment:
Never exercised by tests.

Reply:
Same as above: I don't know how to enforce that `size` has no upper bound.